src/dune_pkg/lock_dir.mli

```diff
@@ -53,6 +53,14 @@ val metadata : Filename.t
 
 module Metadata : Dune_sexp.Versioned_file.S with type data := unit
 
-val write_disk : lock_dir_path:Path.Source.t -> t -> unit
+(** [write_disk_thunk ~lock_dir_path t] returns a thunk that when executed will
+    materialize [t] to disk at the path [lock_dir_path]. If [lock_dir_path]
+    already exists it is first deleted, provided it looks like an existing
+    lockdir. If it doesn't look like a valid lockdir an exception is raised
+    _before_ the thunk is constructed (ie. [write_disk_thunk ~lock_dir_path t]
+    will raise without the unit argument that causes the thunk to execute). This
+    allows callers to separate the validation of the existing lockdir from the
+    side effect that removes the existing lockdir and creates the new one. *)
+val write_disk_thunk : lock_dir_path:Path.Source.t -> t -> unit -> unit
```

Comment:
We usually use `Stage.t` for this idiom.

However, even with `Stage.t` I find this API to be confusing. I think you can just split this into functions:

```
type write
val prepare : .. -> write
val commit : write -> unit
```

Reply:
We usually use `Staged.t` for optimization. E.g. `val find : corpus -> (query -> result) Staged.t`. But you're using to delay a side-effect. It's hard to tell what's the actual side effect that was done immediately versus the one that was delayed to the thunk.